### PR TITLE
feat(aws-ecs-patterns): Service Addon builder pattern

### PR DESCRIPTION
**Pull request by nathanpeck**
This adds a new service builder class to aws-ecs-patterns, which is designed to function using decoupled addons which are additive, rather than large monolithic feature classes. This will make it easier to scale the interface and add native integrations for the growing suite of features. Included in this PR are ECS addons for:

* App Mesh
* FireLens
* X-Ray
* CloudWatch Agent
* Application Load Balancer

This is not yet ready to merge. I want to rename some top level classes to follow the naming convention's of AWS Copilot, and I need to add tests for all the addons. Opening this PR to check if the build will pass remotely as well as locally

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by nathanpeck** [comment]
Tests currently failing due to coverage requirements.

```
@aws-cdk/aws-ecs-patterns: ERROR: Coverage for statements (74.11%) does not meet global threshold (80%)
@aws-cdk/aws-ecs-patterns: =============================== Coverage summary ===============================
@aws-cdk/aws-ecs-patterns: Statements   : 74.11% ( 544/734 )
@aws-cdk/aws-ecs-patterns: Branches     : 71.83% ( 362/504 )
@aws-cdk/aws-ecs-patterns: Functions    : 52.29% ( 57/109 )
@aws-cdk/aws-ecs-patterns: Lines        : 74.11% ( 544/734 )
@aws-cdk/aws-ecs-patterns: ================================================================================
```

I'm working on adding the tests for these new addons now

**Comment by nathanpeck** [comment]
About half of the tests are now added, still some more to come

**Comment by nathanpeck** [comment]
Not quite sure why the integration test I added isn't matching expectations. It looks like it is the same thing just with slightly different VPC cidr blocks, and subnet names. I'll figure out the integration test mismatch tomorrow, but I already verified that the resulting stack deploys and functions on my account via a `cdk deploy` on my test account

**Comment by nathanpeck** [comment]
I realized I was missing a test case for `Service.connectTo` and specifically connecting two App Mesh enabled services to each other. I added this missing test case in the app mesh tests as well as in the integration test

**Comment by nathanpeck** [comment]
@SoManyHs I just added a design doc listing out the goals of this feature and explaining the rationale behind the API design. I totally agree with you that I should have had such a doc from the start, so hopefully this serves to help answer some of your overall concerns on the PR: https://github.com/aws/aws-cdk/blob/51b229834c4d758fee97cbb150a32bfe52b869af/design/aws-ecs/aws-ecs-service-addons.md

Let me know what you think about this and if you agree with the approach described. I'll start addressing some of the other comments as well, but I want to make sure you and the team are all good with the higher level intent behind this PR before I focus too much on the small stuff

File: design/aws-ecs/aws-ecs-service-addons.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+A combination of these hooks should be able to implement any type of functionality that needs to be attached to an ECS service. If we identify new use cases that aren't served by this collection of hooks we can add new hooks to enable new extension types without breaking existing extensions.
+
+## Environments
+
+As part of this proposal there is also a new construct called `Environment`. This construct is designed to be similar to AWS Copilot environments. The intention is that an environment is a transparent wrapper for a VPC, cluster, and any other resources that the environment needs. The environment understands whether it is using EC2 or Fargate capacity. A `Service` is added to an `Environment`. The `Service` changes to EC2 mode if the environment is set to have EC2 capacity. The `Service` changes to Fargate made if the environment is set to have Fargate capacity.
```
Comment (SoManyHs) [inline design/aws-ecs/aws-ecs-service-addons.md:92]:
```suggestion
As part of this proposal there is also a new construct called `Environment`. This construct is designed to be similar to [AWS Copilot environments](https://github.com/aws/copilot-cli). The intention is that an environment is a transparent wrapper for a VPC, cluster, and any other resources that the environment needs. The environment understands whether it is using EC2 or Fargate capacity. A `Service` is added to an `Environment`. The `Service` changes to EC2 mode if the environment is set to have EC2 capacity. The `Service` changes to Fargate mode if the environment is set to have Fargate capacity.
```

File: design/aws-ecs/aws-ecs-service-addons.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+- "cost optimized"
+- "performance optimized"
+- "high availability"
+
+The service would then self configure based on the traits that the customer selected. For example if the service had the "cost optimized" trait for a development environment it might choose to deploy the smallest possible AWS Fargate task size, and only one of them. However the "performance optimized" or "high availability" traits would override this to deploy multiple copies that are larger.
+
+Each extension would be responsible for implementing traits however it choses. For example the FireLens logging extension might decide to implement "cost optimized" by setting a TTL on log files so that they don't accumulate and cost money for storage. However if there was a "data retention" trait that would instead cause FireLens to retain all logs forever for auditing.
+
+This "traits" feature needs more thought before implementation, but it has potential to make it much easier for customer's to define their intention for behavior without needing to explicitly set lower level settings.
```
Comment (SoManyHs) [inline design/aws-ecs/aws-ecs-service-addons.md:108]: I assume this is out of scope for the current design?

File: design/aws-ecs/aws-ecs-service-addons.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+- __Extensions are aware of each other__: There are no hidden "gotchas" when enabling a service extension. Extensions know of each other's existence. For example the application's container is aware that there is an Envoy proxy and that it must wait until the Envoy proxy has started before the app starts.
+- __Service class is agnostic to EC2 vs Fargate__: Rather than choosing up the capacity strategy in the class name customers just create a service, and add it to an environment. The service automatically modifies its settings as needed to adjust to the capacity available in the environment.
+
+
+## Implementation
+
+The new `Service` class acheives its goals via a method `Service.add()` This method is used to add extensions to the service. Each extension is responsible for customizing the service to support a specific feature. The following example shows how the end user instantiates a service with various feature's enabled:
```
Comment (SoManyHs) [inline design/aws-ecs/aws-ecs-service-addons.md:45]:
```suggestion
The new `Service` class achieves its goals via a method `Service.add()` This method is used to add extensions to the service. Each extension is responsible for customizing the service to support a specific feature. The following example shows how the end user instantiates a service with various features enabled:
```

Comment (SoManyHs) [reply design/aws-ecs/aws-ecs-service-addons.md:45]: Are we calling this "extension" or "addon" then?

File: design/aws-ecs/aws-ecs-service-addons.md [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+  memoryMiB: 2048,
+  trafficPort: 80,
+  image: ContainerImage.fromRegistry('nathanpeck/greeter')
+}));
+myService.add(new AppMeshAddon({ mesh }));
+myService.add(new FireLensAddon());
+myService.add(new XRayAddon());
+myService.add(new CloudwatchAgentAddon());
+myService.add(new HttpLoadBalancerAddon());
+```
```
Comment (SoManyHs) [inline design/aws-ecs/aws-ecs-service-addons.md:60]: I love how streamlined this API is, and how we encapsulate extension logic into their own classes. However, do you anticipate certain extensions conflicting with each other? Given that ECS patterns are meant to push best practices in application infrastructure/architecture, how will we provide that guidance with extensions?

File: design/aws-ecs/aws-ecs-service-addons.md [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+Each addon is responsible for one aspect of the resulting service. The final generated service is the sum of all the features that were enabled. There are a few benefits to this approach:
+
+- Users no longer have to select their full architecture up front as with the current ECS patterns. Instead of chosing between `ApplicationLoadBalancedEc2Service` and `NetworkLoadBalancedEc2Service` up front, they can actually deploy a service, and then later add or remove an `HttpLoadBalancerAddon` from the service without needing to change the base class in their code.
+- Users can enable any number of different combinations of features which were not accessible before. They can build the same load balanced services they could before, but now they can also enable App Mesh, X-Ray, Cloudwatch Agent, and FireLens on the service at the same time.
+- The code for each extension is its own self contained class. Rather than bloating the ECS constructs with logic that is specific to another service all of that logic can live in the extension itself.
```
Comment (SoManyHs) [inline design/aws-ecs/aws-ecs-service-addons.md:66]: Again, I like the separation of concerns, but the argument about trying to avoid having a new method on the existing TaskDefinition or Service construct for each new feature, or having a whole new (inflexible) pattern for each new feature in a way is just deferred into a new extension, no?

File: design/aws-ecs/aws-ecs-service-addons.md [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+myService.add(new FireLensAddon());
+myService.add(new XRayAddon());
+myService.add(new CloudwatchAgentAddon());
+myService.add(new HttpLoadBalancerAddon());
+```
+
+Each addon is responsible for one aspect of the resulting service. The final generated service is the sum of all the features that were enabled. There are a few benefits to this approach:
+
+- Users no longer have to select their full architecture up front as with the current ECS patterns. Instead of chosing between `ApplicationLoadBalancedEc2Service` and `NetworkLoadBalancedEc2Service` up front, they can actually deploy a service, and then later add or remove an `HttpLoadBalancerAddon` from the service without needing to change the base class in their code.
+- Users can enable any number of different combinations of features which were not accessible before. They can build the same load balanced services they could before, but now they can also enable App Mesh, X-Ray, Cloudwatch Agent, and FireLens on the service at the same time.
```
Comment (SoManyHs) [inline design/aws-ecs/aws-ecs-service-addons.md:65]: Are there any considerations we need to account for in terms of resource dependency? I see that there's a `bakeContainerDefinition` method that handles container dependencies, but just wondering if it might be necessary to dig into the specific add-ons/extensions proposed here to see if there's anything that might cause a hanging CFN deployment or something like that.

File: design/aws-ecs/aws-ecs-service-addons.md [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+- The code for each extension is its own self contained class. Rather than bloating the ECS constructs with logic that is specific to another service all of that logic can live in the extension itself.
+
+## Service Lifecycle
+
+The new `Service` construct differs a bit from other CDK L2+ constructs. Most CDK constructs are fully assembled in the constructor and are then immutable. The `Service` class is not like this. The constructor does nothing but initialize an empty list of extensions. Users add extensions to this list, to define their intention about what should be built later on. The service constructs do not actually exist at this point. The only thing that exists is a list of user specified intentions about what constructs they do want to exist.
```
Comment (SoManyHs) [inline design/aws-ecs/aws-ecs-service-addons.md:70]: I'd love to hear from @eladb, @RomainMuller  or @rix0rrr or anyone from the core team about using this pattern for constructing a resource and what implications it might have in the JSII.

File: design/aws-ecs/aws-ecs-service-addons.md [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+- `mutateTaskDefinitionProps()` - This is a mutating hook which is passed the `ecs.TaskDefinitionProps` for a TaskDefinition. This allows the extension to make modifications to the task definition, such as App Mesh extension modifying the proxy settings for the task.
+- `useTaskDefinition()` - After the task definition is created, this hook is passed the actual `TaskDefinition` construct that was created. This allows the extension to add containers to the task, modify the task definition's IAM role, etc.
+- `bakeContainerDependencies()` - Once all extensions have added their containers each addon is given a chance to modify its container's `dependsOn` settings. Extensions need to check and see what other extensions were enabled, and decide whether their container needs to wait on another container to start first.
```
Comment (SoManyHs) [inline design/aws-ecs/aws-ecs-service-addons.md:83]: Would `resolveContainerDependencies` be a better descriptor here?

File: design/aws-ecs/aws-ecs-service-addons.md [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+The hooks for an extension are as follows:
+
+- `addHooks()` - This hook is called after all the extensions are added, but before anything else has been run. It gives each extension a chance to do some inspection of the overall service and see what other extensions have been added. Some addons may want to register hooks on the other addons to modify them. For example the Firelens addon wants to be able to modify the settings of the application container to route logs through Firelens.
+- `mutateTaskDefinitionProps()` - This is a mutating hook which is passed the `ecs.TaskDefinitionProps` for a TaskDefinition. This allows the extension to make modifications to the task definition, such as App Mesh extension modifying the proxy settings for the task.
```
Comment (SoManyHs) [inline design/aws-ecs/aws-ecs-service-addons.md:81]: I don't love the verb `mutate` here -- has the Task Definition even been created yet? Would `override` or `modify` work better?

File: design/aws-ecs/aws-ecs-service-addons.md [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+- `useTaskDefinition()` - After the task definition is created, this hook is passed the actual `TaskDefinition` construct that was created. This allows the extension to add containers to the task, modify the task definition's IAM role, etc.
```
Comment (SoManyHs) [inline design/aws-ecs/aws-ecs-service-addons.md:82]: I also feel that `use` is a bit too fuzzy -- maybe `attachTo`? `connectTaskDefinition`?

File: design/aws-ecs/aws-ecs-service-addons.md [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+- `bakeContainerDependencies()` - Once all extensions have added their containers each addon is given a chance to modify its container's `dependsOn` settings. Extensions need to check and see what other extensions were enabled, and decide whether their container needs to wait on another container to start first.
+- `mutateServiceProps()` - Before a service is created this mutating hook is passed a draft service properties to change. For example the App Mesh extension needs to modify the service settings to enable CloudMap service discovery.
+- `useService()` - After the service is created this hook is given a chance to utilize the service that was created. This is used by extensions like the load balancer or App Mesh extension which create and link other AWS resources to the ECS extension.
```
Comment (SoManyHs) [inline design/aws-ecs/aws-ecs-service-addons.md:85]: Similar here

File: design/aws-ecs/aws-ecs-service-addons.md [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+- `connectToService()` - This hook is called when a user wants to connect one service to another service. It allows an extension to implement logic about how to allow connections from one service to another. For example the App Mesh addon implements this method so that you can easily connect one service mesh service to another so that the service's Envoy proxy sidecars know how to route traffic to each other.
```
Comment (SoManyHs) [inline design/aws-ecs/aws-ecs-service-addons.md:86]: Are there any similarities in the connection mechanism? I'm wondering if the API here can kind of model the existing `connections` method.

File: design/aws-ecs/aws-ecs-service-addons.md (hunk 4 above)
Comment (nathanpeck) [reply design/aws-ecs/aws-ecs-service-addons.md:60]: Adding these two questions to the doc for discussion

File: design/aws-ecs/aws-ecs-service-addons.md (hunk 6 above)
Comment (nathanpeck) [reply design/aws-ecs/aws-ecs-service-addons.md:65]: I think with the existing extensions in this PR I have done a lot of testing to make sure that they work out of the box. It is possible that an unhealthy sidecar can cause the task to fail healthchecks and be considered unhealthy, causing a CFN deployment to hang. This shouldn't happen though, and if it does that would be a bug that we fix to make sure that the extension is stable.

File: design/aws-ecs/aws-ecs-service-addons.md (hunk 9 above)
Comment (nathanpeck) [reply design/aws-ecs/aws-ecs-service-addons.md:81]: Sure thing I don't have any issues with changing it to "modify". To be clear this hook only modifies the `props` for the `TaskDefinition` that is about to be created. The `TaskDefinition` has not yet been created at this point, there is just a proposed `props` for it. Each extension is given the option of changing the `props` and then once they have all changed the `props` to fit their needs the resulting `props` object is passed to the `TaskDefinition` constructor to create a new `TaskDefinition`. I'll try to clarify this behavior in the doc

File: design/aws-ecs/aws-ecs-service-addons.md (hunk 10 above)
Comment (nathanpeck) [reply design/aws-ecs/aws-ecs-service-addons.md:82]: I don't think either `attach` or `connect` feels right to me. The purpose of the hook is to let the extension call methods on the created task definition, specifically methods like `TaskDefinition.addContainer()` but also to access `TaskDefinition.taskRole`

I didn't want to make it too specific like `addContainer` because not every extension will want to add a sidecar container. What do you think about switching to a model where the event is less "verb" and more event based, like the hook is called something like `taskDefinitionReady`

File: design/aws-ecs/aws-ecs-service-addons.md (hunk 8 above)
Comment (nathanpeck) [reply design/aws-ecs/aws-ecs-service-addons.md:83]: Sure that sounds good to me

File: design/aws-ecs/aws-ecs-service-addons.md (hunk 2 above)
Comment (nathanpeck) [reply design/aws-ecs/aws-ecs-service-addons.md:108]: Yeah I had it as a stretch goal, but I think given the broad scope of what is already in the PR it would be best to move this stretch goal to its own future enhancement once we get this out in preview already for customers to play with and give feedback

**Comment by nathanpeck** [comment]
Closing this branch in favor of an alternate branch based on the RFC

**Comment by SoManyHs** [comment]
See: https://github.com/aws/aws-cdk/pull/10129
